kernel/CMakeLists.txt

```diff
@@ -115,5 +115,6 @@ target_link_libraries(kernel zephyr_interface)
 endif()
 
 add_dependencies(kernel zephyr_generated_headers)
+add_dependencies(kernel version_h)
```

Comment:
I see that zephyr libraries such as "app" are currently using version.h. These will be broken I suspect.

But having zephyr_generated_headers depend on version_h should be OK.

Reply:
> I see that zephyr libraries such as "app" are currently using version.h. These will be broken I suspect.

I don't see app depends on version.h, where do you see that ?
Could you give a link.

What I do see are bluetooth and net, and a coupld more depending on `version.h`:
https://github.com/zephyrproject-rtos/zephyr/blob/9e4968bb901c8dddde661b8f2d0a0046b338a8ab/subsys/bluetooth/controller/hci/hci.c#L11
https://github.com/zephyrproject-rtos/zephyr/blob/9e4968bb901c8dddde661b8f2d0a0046b338a8ab/subsys/net/l2/openthread/openthread.c#L20

> But having zephyr_generated_headers depend on version_h should be OK.

It should be ok.

Remember, normal source file dependencies will ensure that source files are re-compiled when any of their include headers changes.

What we need to ensure, is that `version.h` are generated (and updated) before the first file included version.h in compiled.
This was why I used `kernel` lib, as that had a direct dependency to `BUILD_VERSION` due to `banner.c`.
Remember, a dependency on `version_h` target doesn't mean it will regenerate the header, only that this target is ensured to be brought up-to-date before compiling the dependent target.

Will check up on the generated headers and update if that is safe, thanks for the suggestion.

Reply:
/home/sebo/ncs/zephyr/samples/subsys/shell/shell_module/src/main.c
9:#include <version.h>

Reply:
ok, a specific sample.
I thought you meant in general.

Reply:
I checked up, the `version_h` target is now a dependency to `zephyr_generated_headers` as I deem this to be safe.

All Zephyr libraries depends on zephyr_generated_headers, and so does kernel lib.

This means the only risk left is custom downstream pure CMake libraries which may not have this dependency specified.
Any downstream pure CMake library depending on `version.h` should create a dependency to either `zephyr_generated_headers` or `version_h` target, if they are not already doing so, to be on the safe side.

But as all Zephyr libraries are already depending on `zephyr_generated_headers`, then in practice, `version.h` will be one of the first build steps to be carried out, so there is only a very small risk if a downstream library forgets to do this.